### DISCUSSION TITLE
Extending username password form Authentication

### DISCUSSION THREAD

**Discussion by Bhavik237** [keycloak/keycloak#20987] [Archive]
Hello everyone,

I am currently working on extending the Keycloak Authenticator's username-password form, and I need some guidance on implementing a specific functionality. My goal is to enhance the authentication process by incorporating a default authentication method that is attempted first. If the default authentication fails, I would like to seamlessly transition to a custom authentication method. I believe this approach will provide a more robust and flexible authentication experience. I would appreciate any advice or insights on how to accomplish this extension in Keycloak. Thank you in advance for your help!

I am using this interface but i am getting "| Caused by: java.lang.ClassNotFoundException: org.keycloak.authentication.DisplayTypeAuthenticatorFactory
"
`public class UsernamePasswordAttributeFormFactory implements AuthenticatorFactory, DisplayTypeAuthenticatorFactory `
`public class UsernamePasswordAttributeForm extends UsernamePasswordForm implements Authenticator`

is my approach correct ? or should I try something else

Thanks

**Comment by danielFesenmeyer** [comment]
The `DisplayTypeAuthenticatorFactory` has been removed in one of the last keycloak releases.

It seems that the keycloak version you are using for implementation does not match the one you are using to run your authenticator. You should make sure that those versions match.

**Reply by Bhavik237** [reply]
yes I have change it now. it is working fine. but i got another problem ,can you please have look

**Comment by Bhavik237** [comment]
![untitled (1)](https://github.com/keycloak/keycloak/assets/57186072/7efd64c8-dcd0-4171-a19f-aac386c57e83)

provide image i want to achieve . when i validate from external link I get following error . (external service is giving response as true)
```
testkeycloak    | 2023-06-14 12:06:15,846 WARN  [org.keycloak.services] (executor-thread-35) KC-SERVICES0013: Failed authentication: org.keycloak.authentication.AuthenticationFlowException
testkeycloak    | 	at org.keycloak.authentication.DefaultAuthenticationFlow.processResult(DefaultAuthenticationFlow.java:497)
testkeycloak    | 	at org.keycloak.authentication.DefaultAuthenticationFlow.processAction(DefaultAuthenticationFlow.java:156)
testkeycloak    | 	at org.keycloak.authentication.AuthenticationProcessor.authenticationAction(AuthenticationProcessor.java:977)
testkeycloak    | 	at org.keycloak.services.resources.LoginActionsService.processFlow(LoginActionsService.java:311)
testkeycloak    | 	at org.keycloak.services.resources.LoginActionsService.processAuthentication(LoginActionsService.java:282)
testkeycloak    | 	at org.keycloak.services.resources.LoginActionsService.authenticate(LoginActionsService.java:274)
testkeycloak    | 	at org.keycloak.services.resources.LoginActionsService.authenticateForm(LoginActionsService.java:339)
testkeycloak    | 	at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
testkeycloak    | 	at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
testkeycloak    | 	at java.base/jdk.internal.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
testkeycloak    | 	at java.base/java.lang.reflect.Method.invoke(Method.java:566)
testkeycloak    | 	at org.jboss.resteasy.core.MethodInjectorImpl.invoke(MethodInjectorImpl.java:170)
testkeycloak    | 	at org.jboss.resteasy.core.MethodInjectorImpl.invoke(MethodInjectorImpl.java:130)
testkeycloak    | 	at org.jboss.resteasy.core.ResourceMethodInvoker.internalInvokeOnTarget(ResourceMethodInvoker.java:660)
testkeycloak    | 	at org.jboss.resteasy.core.ResourceMethodInvoker.invokeOnTargetAfterFilter(ResourceMethodInvoker.java:524)
testkeycloak    | 	at org.jboss.resteasy.core.ResourceMethodInvoker.lambda$invokeOnTarget$2(ResourceMethodInvoker.java:474)
testkeycloak    | 	at org.jboss.resteasy.core.interception.jaxrs.PreMatchContainerRequestContext.filter(PreMatchContainerRequestContext.java:364)
testkeycloak    | 	at org.jboss.resteasy.core.ResourceMethodInvoker.invokeOnTarget(ResourceMethodInvoker.java:476)


the code snippet 
```
public class UsernamePasswordAttributeForm extends UsernamePasswordForm implements Authenticator{
    protected static ServicesLogger log = ServicesLogger.LOGGER;
    private final HttpClient client;
    ObjectMapper objectMapper = new ObjectMapper();

    public UsernamePasswordAttributeForm(HttpClient client) {
        this.client = client;
    }

    @Override
    public void action(AuthenticationFlowContext context) {
        MultivaluedMap<String, String> formData = context.getHttpRequest().getDecodedFormParameters();

        if (formData.containsKey("cancel")) {
            log.info("alternative-username-password-form : Form was canceled.");
            context.cancelLogin();
            return;
        }
        //!validateForm(context, formData) ||
        if (!checkUserInExternalService(context)) {
            log.info("B :"+validateForm(context, formData)+"A :"+checkUserInExternalService(context)+"  " );
            log.info("alternative-username-password-form : Oops! User is not valid : wrong credentials or unknown");

            boolean isRequired = context.getExecution().isRequired();
            if (!isRequired) {
                log.info("alternative-username-password-form : Execution is not required. Continuing the execution flow to the next 'alternative' task.");
                context.attempted();
                return;
            }

            log.info("alternative-username-password-form : Resetting the flow, all stops here.");
            context.getEvent().error(Errors.USER_NOT_FOUND);
            context.failure(AuthenticationFlowError.UNKNOWN_USER);
            return;

        }
        log.info("B :"+validateForm(context, formData)+" A :"+checkUserInExternalService(context)+"  " );
        log.info("alternative-username-password-form : User is valid and authenticated.");
        context.success();
    }


flow changes:
![Screenshot from 2023-06-14 18-03-29](https://github.com/keycloak/keycloak/assets/57186072/0cd7787b-00c5-43ed-8b60-b329916234b4)
```
```

**Reply by danielFesenmeyer** [reply]
Probably it makes sense that you debug that code.

That warning appears in the `failure`-scenario, right?
I guess the reason for that error is that you haven't added a challenge to the context. In this case, the exception is thrown, which contains your provided error code AuthenticationFlow.UNKNOWN_USER - but you don't see that in the log message.
In the DefaultAuthenticationFlow class (directly above where the exception is thrown), you see that failures need to have a challenge - otherwise, I think a generic error message is shown.

**Reply by nowilliemotley** [reply]
Nice recomendation

**Reply by Bhavik237** [reply]
okay I will try it.

**Reply by Bhavik237** [reply]
Hi @danielFesenmeyer can we connect if possible I am not able to solve .I am pretty new to custom provider.

**Reply by danielFesenmeyer** [reply]
Hi @Bhavik237, I'm sorry, I don't have much time in the next days. I hope you can figure it out or someone else can assist.

**Reply by Bhavik237** [reply]
okay @danielFesenmeyer , Thanks for the help.

**Reply by vidurgupta-fareye** [reply]
@Bhavik237 Did you solve this, can you share the complete extension class? I am in the process of upgrading my keycloak from 18 to 23.

**Reply by Bhavik237** [reply]
@vidurgupta-fareye can you give exact class name that you want ?

**Reply by vidurgupta-fareye** [reply]
@Bhavik237 The one that was earlier implementing DisplayTypeAuthenticatorFactory. How did you achieve the same in later version,

**Reply by kiran-sg** [reply]
@Bhavik237 have you solved this issue? Can you please share your custom class which was having the issue.
